odoo/cli/shell.py

```diff
@@ -1,159 +1,173 @@
-----------------------------------------------------------------------
- python   | stop  | raise  | raise  | raise      | raise
- ipython  | stop  | stop   | stop   | loop       | loop
- ptpython | stop  | raise  | raise  | raise      | raise
- bpython  | stop  | stop   | stop   | stop       | stop
+    shell    | ^D    | exit() | quit() | sys.exit() | raise SystemExit()
+    ----------------------------------------------------------------------
+    python   | stop  | raise  | raise  | raise      | raise
+    ipython  | stop  | stop   | stop   | loop       | loop
+    ptpython | stop  | raise  | raise  | raise      | raise
+    bpython  | stop  | stop   | stop   | stop       | stop
 
+    stop = The REPL main loop stop.
+    raise = Exception raised.
+    loop = Stay in REPL.
 """
 
 
-def raise_keyboard_interrupt(*a):
-    raise KeyboardInterrupt()
+class Shell(Command):
+    """Start odoo in an interactive shell"""
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.parser.add_argument(
+            '-c', '--config', dest='config',
+            help="use a specific configuration file")
+        self.parser.add_argument(
+            '-d', '--database', dest='db_name', default=None,
+            help="database name, connection details will be taken from the config file")
+        self.parser.add_argument(
+            '--shell-file', default=os.environ.get('PYTHONSTARTUP'),
+            help="Specify a python script to be run after the start of the shell. "
+                 "Overrides the env variable PYTHONSTARTUP.")
+        self.parser.add_argument(
+            '--shell-interface', choices=SHELLS,
+            help="Specify a preferred REPL to use in shell mode.")
 
-class Console(code.InteractiveConsole):
-    def __init__(self, local_vars=None, filename="<console>"):
-        code.InteractiveConsole.__init__(self, locals=local_vars, filename=filename)
-        try:
-            import readline
-            import rlcompleter
-        except ImportError:
-            print('readline or rlcompleter not available, autocomplete disabled.')
+    @contextmanager
+    def _build_env(self, dbname):
+        if dbname:
+            with Registry(dbname).cursor() as cr:
+                ctx = Environment(cr, SUPERUSER_ID, {})['res.users'].context_get()
+                env = Environment(cr, SUPERUSER_ID, ctx)
+                yield env
+                cr.rollback()
         else:
-            readline.set_completer(rlcompleter.Completer(local_vars).complete)
-            readline.parse_and_bind("tab: complete")
-
+            yield None
 
-class Shell(Command):
-    """Start odoo in an interactive shell"""
-    supported_shells = ['ipython', 'ptpython', 'bpython', 'python']
+    def run(self, cmdargs):
+        parsed_args = self.parser.parse_args(args=cmdargs)
 
-    def init(self, args):
-        config.parser.prog = self.prog
```

Comment:
optparse vs argparse

I know that the former is deprecated, but we have tons of useful options there: logging, database connnection settings, etc. My comment is similar as to the one for the `test` subcommand. We need a lot of these options.

Reply:
I mean this is breaking usage such as:
`odoo-bin shell --log-level debug_sql`